[MM-15887] E2E: add test for interactive menus - basic options

saturninoabril opened the pull request:
#### Summary
Add E2E for interactive menus - basic options.  This PR covered test cases for the ff:
![Screen Shot 2019-07-06 at 1 04 14 AM](https://user-images.githubusercontent.com/5334504/60736607-1fd70800-9f8a-11e9-8d76-059280368730.png)

__To follow: Test cases for interactive menus with users and channels as options.__

Other changes included:
- add small webhook server with additional `express` and `body-parser` dependencies on `/e2e` subproject.  Such webhook server is exposed as `http://localhost:3000` on local dev and will be `https://cypress.test.masttermost.com/webhook` on Cypress test server (PR to follow on shared pipelines)
- introduced `NODE_ENV=qa` which will be used when running test at Cypress test server. Such will be used to configure other settings like `baseUrl`, `mailboxAPI`, etc. (PR to follow both on the webapp and shared pipelines)

#### Ticket Link
Jira ticket: [MM-15887](https://mattermost.atlassian.net/browse/MM-15887)

saturninoabril commented on e2e/cypress/fixtures/hooks/message_menus_with_datasource.json:
```diff
@@ -0,0 +1,17 @@
+            "data_source": "channels"
+        }]
+    }]
+}
```
Above payloads are the same as the ones we have at https://docs.mattermost.com/developer/interactive-messages.html#message-menus

saturninoabril commented on e2e/cypress/integration/interactive_menu/basic_options.js:
```diff
@@ -0,0 +1,159 @@
+// ***************************************************************
+
+/**
+* Note: This test requires webhook server running. Initiate `npm run start:webhook` to start.
```
There's a module that we can use to [start a server](https://docs.cypress.io/guides/guides/continuous-integration.html#Boot-your-server) before doing test but I initially thought that it's not necessary since not all specs/files are using that webhook server.  So I just put a note only as a reminder for local development.  On test server, this will be started via CLI command (PR to follow in shared pipelines).

thekiiingbob commented on e2e/cypress/integration/interactive_menu/basic_options.js:
```diff
@@ -0,0 +1,159 @@
+    it('matches elements', () => {
+        // # Post an incoming webhook
+        cy.task('postIncomingWebhook', {url: incomingWebhook.url, data: payload});
+        cy.wait(TIMEOUTS.TINY);
```
You might be able to omit this wait if you use an assertion off of the task. Maybe just `should('have.property', 'body')` could be sufficient enough? (It looks like there are a few other instances of this in this file as well).

It is just a tiny wait though, so I'll approve the PR and you can let me know if it's something that you feel is worth doing.

saturninoabril replied:
Thanks Bob, I forgot about it and it should be it.  I'll update the PR.

thekiiingbob commented:
@saturninoabril  just had one question/suggestion about some waits.

saturninoabril commented:
@thekiiingbob I've added status check to `postIncomingWebhook` and did the same for `postMessageAs`.  Also fixed the way it returns data from those task.  Please take a look again.

For other `postMessageAs` instances, I'll send separate PR to have similar fix on how we're using it.  I think some of the remaining flaky test attributes to it (should wait for the status to be checked before doing next test step).

saturninoabril commented on e2e/cypress/integration/interactive_menu/basic_options.js:
```diff
@@ -106,7 +108,7 @@ describe('MM-15887 Interactive menus - basic options', () => {
         });
 
-        cy.wait(TIMEOUTS.TINY);
+        cy.wait(TIMEOUTS.SMALL);
```
I increased the `wait` here to ensure that it'll assert the ephemeral post after clicking the option.  Can't think of a way to better handle it.

thekiiingbob replied:
Yeah, that's tricky. I feel like we have the same issue for knowing when a new post arrives. We could do something like track the last post ID, and wait for a new one, but that feels like a lot of overheard vs  "hey, just wait a second".
